lib/ret/api/credentials.ex

```diff
@@ -0,0 +1,115 @@
+defmodule Ret.Api.Credentials do
+  @moduledoc """
+  Credentials for API access.
+  """
+  alias Ret.Api.Credentials
+
+  alias Ret.Account
+
+  use Ecto.Schema
+  import Ecto.Query, only: [from: 2]
+  import Ecto.Changeset
+
+  @schema_prefix "ret0"
+  @primary_key {:api_credentials_id, :id, autogenerate: true}
+
+  schema "api_credentials" do
+    field(:api_credentials_sid, :string)
+    field(:token_hash, :string)
+    field(:subject_type, Ret.Api.TokenSubjectType)
+    field(:issued_at, :utc_datetime)
+    field(:expires_at, :utc_datetime)
+    field(:is_revoked, :boolean)
+    field(:scopes, {:array, Ret.Api.ScopeType})
+
+    belongs_to(:account, Account, references: :account_id)
+    timestamps()
+  end
+
+  def create_new_api_credentials(options) do
+    %Credentials{}
+    |> changeset(options)
+    |> Ret.Repo.insert()
+  end
+
+  defp changeset(%Credentials{} = credentials, %{
+         token_hash: token_hash,
+         subject_type: subject_type,
+         scopes: scopes,
+         account: account
+       }) do
+    credentials
+    |> change()
+    |> add_api_credentials_sid_to_changeset
+    |> add_token_hash_to_changeset(token_hash)
+    |> add_subject_type_to_changeset(subject_type)
+    |> add_issued_at_to_changeset(Timex.now() |> DateTime.truncate(:second))
```

Comment:
Initially it was just monkey-see monkey-do
```elixir
./lib/ret/web_push_subscription.ex:95:    |> put_change(:last_notified_at, Timex.now() |> DateTime.truncate(:second))
./lib/ret/hub.ex:297:    do: changeset |> put_change(:last_active_at, Timex.now() |> DateTime.truncate(:second))
./lib/ret/hub.ex:504:    expires_at = Timex.now() |> Timex.shift(hours: @entry_code_expiration_hours) |> DateTime.truncate(:second)
./lib/ret/scene.ex:245:    |> put_change(:reviewed_at, Timex.now() |> DateTime.truncate(:second))
./lib/ret/jobs/stats_job.ex:14:             measured_at <- NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second),
./lib/ret/account_favorite.ex:65:    |> put_change(:last_activated_at, Timex.now() |> DateTime.truncate(:second))
./lib/ret/account_favorite.ex:71:    |> put_change(:last_activated_at, Timex.now() |> DateTime.truncate(:second))
./lib/ret_web/channels/session_socket.ex:17:      |> assign(:started_at, NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second))
./lib/ret_web/channels/session_socket.ex:26:      |> assign(:started_at, NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second))
./priv/repo/migrations/20181128022000_add_min_token_issued_at_to_account.exs:5:    epoch = DateTime.from_unix!(0, :second) |> DateTime.truncate(:second) |> DateTime.to_iso8601()
./test/ret/guardian_test.exs:25:    date = Timex.now() |> Timex.shift(seconds: 1) |> DateTime.truncate(:second)
```
but then when I tested without truncating I saw it was required:
```elixir
** (ArgumentError) :utc_datetime expects microseconds to be empty, got: ~U[2021-11-13 00:21:40.626028Z]

Use `DateTime.truncate(utc_datetime, :second)` (available in Elixir v1.6+) to remove microseconds.

    (ecto) lib/ecto/type.ex:1359: Ecto.Type.check_no_usec!/2
    (ecto) lib/ecto/type.ex:487: Ecto.Type.dump_utc_datetime/1
    (ecto) lib/ecto/type.ex:946: Ecto.Type.process_dumpers/3
    (ecto) lib/ecto/repo/schema.ex:930: Ecto.Repo.Schema.dump_field!/6
    (ecto) lib/ecto/repo/schema.ex:943: anonymous fn/6 in Ecto.Repo.Schema.dump_fields!/5
    (stdlib) maps.erl:232: :maps.fold_1/3
    (ecto) lib/ecto/repo/schema.ex:941: Ecto.Repo.Schema.dump_fields!/5
    (ecto) lib/ecto/repo/schema.ex:874: Ecto.Repo.Schema.dump_changes!/6
    (ecto) lib/ecto/repo/schema.ex:256: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
    (ret) lib/ret/api/token_module.ex:59: Ret.Api.TokenModule.create_token/3
    (guardian) lib/guardian.ex:765: Guardian.returning_tuple/1
    (guardian) lib/guardian.ex:594: Guardian.encode_and_sign/4
    (ret) lib/mix/tasks/generate_api_token.ex:59: Mix.Tasks.GenerateApiToken.gen_app_token/0
    (mix) lib/mix/task.ex:331: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:79: Mix.CLI.run_task/2
```
I haven't tracked down exactly when/why microseconds are not accepted. It seems to have been a change in ecto as of 3.0.0:
```
[Ecto.Schema] :time, :naive_datetime and :utc_datetime no longer keep microseconds information. If you want to keep microseconds, use :time_usec, :naive_datetime_usec, :utc_datetime_usec
```
(In the "Backwards Incompatible" section of the [changelog](https://github.com/elixir-ecto/ecto/blob/001ff8f1d4122045a194ba693a555e61eb562ee0/CHANGELOG.md#v300-2018-10-29))

Reply:
Ah, I didn't realize we were already doing this in the rest of the codebase. I guess we could just truncate microseconds if that's the issue. Not a big deal, just seemed weird when I saw it.

Reply:
It looks like we need to truncate at `:seconds`.  "If you want to keep microseconds..." actually means "If you want to keep anything more precise than seconds...", because `check_no_usec!` makes sure the `microseconds: {microseconds, precision}` tuple is `{0,0}`:
[`check_no_usec!`](https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/type.ex#L1308-L1316)
[`truncate`](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/calendar.ex#L347-L361) (called from https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/calendar/datetime.ex#L1375)

Reply:
I see, thanks for digging deeper. Makes sense to leave the `truncate(:second)` in then.